sb3_contrib/trpo/trpo.py

```diff
@@ -0,0 +1,387 @@
+                beta = th.sqrt(beta)
+
+                alpha = 1
+                orig_params = [param.detach().clone() for param in params]
+
+                is_line_search_success = False
+                with th.no_grad():
+                    # Line-search
+                    for _ in range(self.ls_steps):
+
+                        j = 0
+                        # Applying the scaled step direction
+                        for param, orig_param, shape in zip(params, orig_params, grad_shape):
+                            k = param.numel()
+                            param.data = orig_param.data + alpha * beta * search_direction[j : (j + k)].view(shape)
+                            j += k
+
+                        # Recomputing the policy log-probabilities
+                        _, log_prob, _ = self.policy.evaluate_actions(rollout_data.observations, actions)
+
+                        # New policy objective
+                        ratio = th.exp(log_prob - rollout_data.old_log_prob)
+                        new_policy_obj = (advantages * ratio).mean()
+
+                        # New KL-divergence
+                        kl_div = kl_divergence(distribution.distribution, old_distribution.distribution).mean()
+
+                        # Constraint criteria
+                        if (kl_div < self.target_kl) and (new_policy_obj > policy_obj):
+                            is_line_search_success = True
+                            break
+
+                        # Reducing step size if line-search wasn't successful
+                        alpha *= self.ls_alpha
+
+                    line_search_results.append(is_line_search_success)
+
+                    if not is_line_search_success:
+                        # If the line-search wasn't successful we revert to the original parameters
+                        for param, orig_param in zip(params, orig_params):
+                            param.data = orig_param.data.clone()
+
+                        po_values.append(policy_obj.item())
+                        kl_divergences.append(0)
+                    else:
+                        po_values.append(new_policy_obj.item())
+                        kl_divergences.append(kl_div.item())
+
+                # Critic updates
+                for _ in range(self.n_critic_updates):
+                    values, _, _ = self.policy.evaluate_actions(rollout_data.observations, actions)
+                    values_pred = values.flatten()
+                    value_loss = F.mse_loss(rollout_data.returns, values_pred)
+                    value_losses.append(value_loss.item())
+
+                    self.policy.optimizer.zero_grad()
+                    value_loss.backward()
+                    # Removing gradients of parameters shared with the actor
+                    # otherwise it defeats the purposes of the KL constraint
+                    for param in params:
```

Comment:
it seems that you are removing all gradients from params from `self.policy` no? (which contains the critic params too)

Reply:
`params` will only contain parameters which have gradients w.r.t to the KL divergence - for which we don't want to update in the Critic step (shared layers) otherwise it might violate the KL constraint.